src/test/java/edu/hm/hafner/util/SetTest.java

```diff
@@ -0,0 +1,179 @@
+        testHash.remove(500);
+        assertThat(testHash.iterator().hasNext()).isFalse();
+    }
+        
+    @Test
+    void size() {
+        testInitializing();
+
+        //Test: Sizes after declaration.
+        assertThat(testHash.size()).isEqualTo(0);
+        assertThat(extendHash.size()).isEqualTo(4);
+
+        // Test: Adding values
+        testHash.addAll(extendHash);
+        assertThat(testHash.size()).isEqualTo(4);
+        testHash.add(360);
+        assertThat(testHash.size()).isEqualTo(5);
+        testHash.add(360);
+        assertThat(testHash.size()).isEqualTo(5);
+
+        // Test: Size after removing a value.
+        testHash.remove(360);
+        assertThat(testHash.size()).isEqualTo(4);
+        testHash.removeAll(extendHash);
+        assertThat(testHash.size()).isEqualTo(0);
+    }
+
+    @Test
+    void isEmpty() {
+        testInitializing();
+
+        //Test: Creating a new variable
+        assertThat(testHash.isEmpty()).isTrue();
+
+        //Test: Adding values, so the hash is not empty any more.
+        testHash.add(2503);
+        assertThat(testHash.isEmpty()).isFalse();
+        testHash.add(1106);
+        assertThat(testHash.isEmpty()).isFalse();
+        testHash.add(2212);
+        assertThat(testHash.isEmpty()).isFalse();
+        testHash.add(709);
+        assertThat(testHash.isEmpty()).isFalse();
+
+        //Test: Removing all values in the hash step by step.
+        testHash.remove(2503);
+        assertThat(testHash.isEmpty()).isFalse();
+        testHash.remove(1106);
+        assertThat(testHash.isEmpty()).isFalse();
+        testHash.remove(2212);
+        assertThat(testHash.isEmpty()).isFalse();
+        testHash.remove(709);
+        assertThat(testHash.isEmpty()).isTrue();
+
+        // Test: Adding and removing more than one Object.
+        testHash.addAll(extendHash);
+        assertThat(testHash.isEmpty()).isFalse();
+        testHash.removeAll(extendHash);
+        assertThat(testHash.isEmpty()).isTrue();
+    }
+
+    @Test
+    void contains() {
+        testInitializing();
+        assertThat(extendHash.contains(1106)).isTrue();
+        assertThat(extendHash.contains(2503)).isTrue();
+        assertThat(extendHash.contains(2212)).isTrue();
+        assertThat(extendHash.contains(709)).isTrue();
+        assertThat(extendHash.contains(4)).isFalse();
+        assertThat(extendHash.contains(0)).isFalse();
+        assertThat(extendHash.contains(1999)).isFalse();
+    }
+
+    @Test
+    void add() {
+        testHash = new HashSet<>();
+        assertThat(testHash).isEmpty();
+        testHash.add(50);
+        assertThat(testHash.hashCode()).isEqualTo(50);
+        testHash.add(443);
+        assertThat(testHash.hashCode()).isEqualTo(493); //50+443
+    }
+
+    @Test
+    void remove() {
+        testInitializing();
+        assertThat(extendHash).isNotNull();
+        assertThat(extendHash.hashCode()).isEqualTo(6530); //All extendedHash integers summed
```

Comment:
In Tests möglichst nicht mit `hashCode` arbeiten. Der Test und seine Korrektheit sind so nur schwer nachzuvollziehen.

Reply:
Warum ist der Test mit hashCode nicht so gut? Wie kann ich sonst die Methoden richtig testen? Mit hasNext()?

Reply:
Mit `contains` oder `containsAll` könntest du das z.B. testen. Von assertJ gibt es auch Methoden die überprüfen ob bestimmte Elemente enthalten sind. Die wären auch eine Möglichkeit. Dadurch kann man die Tests deutlich besser nachvollziehen. Ich musste zum Beispiel erst mal nachschauen, ob der HashCode eines Sets als Summe definiert ist (Die Definition hat normalerweise keine besondere Bedeutung und ist deswegen unwichtig). Dazu kommt, dass die Summe nicht eindeutig auf die enthaltenen Elemente schließen lässt. Z.B.: [2, 3] => 5 und [1, 4] => 5.